prow/git/git.go

```diff
@@ -288,9 +296,19 @@ func (r *Repo) Push(repo, branch string) error {
 		return errors.New("cannot push without credentials - configure your git client")
 	}
 	r.logger.Infof("Pushing to '%s/%s (branch: %s)'.", r.user, repo, branch)
-	remote := fmt.Sprintf("https://%s:%s@%s/%s/%s", r.user, r.pass, github, r.user, repo)
-	co := r.gitCommand("push", remote, branch)
-	_, err := co.CombinedOutput()
+
+	remote, err := Remote(r.base, r.user, r.pass, r.user, repo)
+	if err != nil {
+		return fmt.Errorf("Remote error: %v", err)
+	}
+
+	// We can't push without user and repo in remote path
+	if !strings.Contains(remote.String(), strings.Join([]string{r.user, repo}, "/")) {
+		return errors.New("git remote must contain :user and :repo to push")
```

Comment:
Shouldn't this be validated at flag parse time? Why bother creating a `git` client that can't push?

Reply:
I can remove this, but I put it in here to prevent weird silent errors, that could be introduced by flag logic changes.

It's up to you guys, leave it in, take it out... I am good either way.

Reply:
Steve what is your concern with doing this validation?

I can see the case for being too lazy to add it, but why ask to remove it?

Reply:
It just seems crazy to me. Like why would you validate this at flag load time, then call a func with it, then sanity-check that the func did what you wanted? It's just muddying the code. We should write code that validates up front and uses types where possible to make it impossible to error later.

Reply:
What's the verdict here, leave it in, or take it out?

Reply:
I believe it would be better for our code health to use the type system here to make sure that if we validate this upfront, it is not possible for the code downstream of it to error, and that would let us not need these runtime sanity checks

Reply:
Deal... let me do a commit.